### ISSUE TITLE
Different column name for jsonb backend

### ISSUE THREAD

**Issue by ryanb** [shioyama/mobility#91]
It would be nice if the `jsonb` and `hstore` backends had a different default column name such as `title_translations` for a `title` attribute. This way we are not overriding methods generated by ActiveRecord and we can access it directly without using a work around like `read_attribute`.

Also `read_attribute` might be used by other gems as an alternative way to access the value and expect it to be the same as calling the accessor method.

The real motivation for this change is that we can define `type_for_attribute` or `column_for_attribute` methods so that other gems can infer the type (such as simple_form and formtastic). Overriding this on a column which already exists might have some bad side effects.

This change would probably require a major version bump so I can understand if it gets rejected.

**Comment by shioyama** [comment]
Yes, I actually made a conscious decision to use the attribute name as the column name. If you look around at other gems that support hstore/jsonb, you'll see that there are examples of both (using a column with a prefix, like `title_translations`, and using the attribute itself, i.e. `title`).

I went with using the attribute itself for two reasons: (1) anything else would be somewhat arbitrary, in that who am I to decide what prefix/postfix we should append to the attribute in the column?, and (2) namespacing concerns that e.g. `title_translations` might conflict with other method names (which in fact is a real concern, since in 0.3.0 I'm planning to replace the `title_backend` methods with `title_translations`).

What I **would** consider would be to have optional `column_prefix` and `column_postfix` options that you could optionally set to customize the choice of column name. This would be fairly easy to implement and would I think satisfy your need.

Also, note that to get the original column, you can use the `super` option (as of 0.2.0). So `post.title(super: true)` will return the column value (a hash with all locale translations). This may or may not solve your problem.

**Comment by ryanb** [comment]
> What I would consider would be to have optional column_prefix and column_postfix options that you could optionally set to customize the choice of column name. This would be fairly easy to implement and would I think satisfy your need.

That would be handy, thanks. The main problem I have with overriding the current method is that we now have a method which returns different types depending on how it is called. This causes confusion with gems such as `simple_form` which infer the type from the column.

Also, I have run into odd edge cases where the JSON value is used instead of the translated value. Such as when there is a validation error for some reason I get this:

<img width="503" alt="screen shot 2017-09-18 at 10 04 24 am" src="https://user-images.githubusercontent.com/161/30554279-dc8a81d4-9c58-11e7-936e-fe251924fe72.png">

I am not yet certain why this is happening with `simple_form` but my current solution is to set the type and value manually:

```erb
<%= f.input(:name, as: :string, input_html: {value: @course.name}) %>
```

These tricky edge cases and type inference would not be an issue if the translated column was a different name. Then `mobility` could truly simulate a database column instead of juggling two different types.

**Comment by ryanb** [comment]
Just noticed another case where this would be useful. Carrierwave [uses read_attribute](https://github.com/carrierwaveuploader/carrierwave/blob/16c0eb5a7869d49930e165bc6a519033f067e9c6/lib/carrierwave/orm/activerecord.rb#L44) so having a file which is different per translation requires overriding methods like this:

```ruby
def read_uploader(column)
  send("#{column}_backend").read(Mobility.locale)
end

def write_uploader(column, value)
  send("#{column}_backend").write(Mobility.locale, value)
end
```

If mobility did override `read_attribute` it would handle this automatically.

**Comment by shioyama** [comment]
Hmm thanks for the food for thought! Especially the idea of using the single `translations` jsonb column, that's a very interesting idea (for onother backend). And the point about simple_form is really important.

About overriding `read_attribute` though, I've thought about this but I feel this is messing too deep with the internals of ActiveRecord. Globalize [does this](https://github.com/globalize/globalize/blob/master/lib/globalize/active_record/instance_methods.rb#L42-L55), and I initially wondered why it was necessary, but I'm now [starting to see why](https://github.com/rmm5t/strip_attributes/pull/34). Still, I'm against it because I think it's "tricking" AR and doing that is a slippery slope -- you start having to do more and more things to make AR/Rails believe that the attribute is a "real" attribute when in fact it's something a bit different.

I'm actually more inclined to do less rather than more. For example, the problem linked to above with `strip_attributes` is caused by the fact that Mobility adds translated attributes to the AR `attributes` hash, but I'm wondering now if this is a good idea, because gems like `strip_attributes` take this as an indication as the attribute should be callable using `read_attribute`, when in fact it is not.

So... I know it's tempting to override those methods, but I think in the longterm (from my experience working on Globalize as well), it's better not to do that and have the user of Mobility be aware that the gem works this way.

That said, your argument about the jsonb column can be handled in different ways, for example having an option and defaulting it to a different column name would be one way to resolve that one.

**Comment by shioyama** [comment]
@ryanb #99 is another issue caused by the fact that Mobility is using a jsonb column with the same name as the translated attribute name. So I am now leaning toward your suggestion to name the column something different, with prefix/suffix options allowing the user to change this to something else if they want.

However, I'm still against overriding `read_attribute` and `write_attribute`. Instead (as a partial fix) I'm thinking of removing [the code](https://github.com/shioyama/mobility/blob/e803847f967117a34ef8f31491af555836d97c6d/lib/mobility/active_record/attribute_methods.rb#L29-L33) which currently adds translated attributes to the `attributes` hash, which is bad because it implies that you should be able to pass these translated attributes to `read_attributes` and get the translation, which you can't.

I think doing both of these should go some way to resolving some (but not all) of the issues mentioned here.

**Comment by ryanb** [comment]
That sounds like a good solution since we can override read/write attribute manually without too much headache if the column is named something different. I wish ActiveRecord had a better defined interface for accessing virtual attributes but oh well.

Thanks for your work on this gem!

**Comment by shioyama** [comment]
For the record, since this is a breaking change, I'm pushing it off until the next major release (i.e. 1.0). We'll release 0.3.0 without it for now.

**Comment by shioyama** [comment]
I've just merged #189 which adds `prefix` and `suffix` options for hash-valued backends like Json/Jsonb/Hstore (also Serialized). Although this doesn't *force* people to use a different column name than the attribute name, it at least now gives them the possibility to do it.

I'm going to close this since I think the "issue" is resolved insofar as it is now resolvable. Once I release a new version I'll add mention to the readme + wiki that it's recommended to add a prefix and/or suffix when using these backends.
